README.md
<p align="center">
  <img src="logo/1024px.png" alt="ubuntu_install" height="200px">
  </br>
  <em> Logo designed by <a href="https://github.com/familqasimov">Famil Qasimov</a> </em>
</p>


ubuntu install
===========
[![environment](https://img.shields.io/badge/linux-ubuntu_18.04-orange.svg)](https://img.shields.io/badge/linux-ubuntu_18.04-orange.svg) [![environment](https://img.shields.io/badge/language-ansible-2196F3.svg)](https://img.shields.io/badge/tool-ansible-2196F3.svg) [![environment](https://img.shields.io/badge/language-shell-43A047.svg)](https://img.shields.io/badge/tool-ansible-2196F3.svg)

## Description
Install my firsts packages and make initial configs in a Ubuntu formatted. Its recommended Ubuntu 18.04 Minimal Installation.

OBS: The most of the installation is done by [Ansible](https://github.com/ansible/ansible)

The installation is divided in 3 parts

> 1. Base

This installation contains a principal packages and only kind of intallattion triggered by [``trigger.sh``](https://raw.githubusercontent.com/frankjuniorr/ubuntu_install/master/code/trigger.sh)

> 2. Optional

This installation contains others kind of packages not so importants to initial setup (ex. Telegram, LibreOffice and Transmission)

> 3. Development

TODO

## Base Installation

### 1. Initial Upgrade
Make the initial config with `sudo apt upgrade && sudo apt dist-upgrade`
### 2. Install packages
Install many packages. Basically the packages are:
- Spotify
- Visual Studio Code
- Some tools like: git, vim, meld, curl, wget and others stuffs
- Dropbox
- Google Chrome
- Terminator
- Some dark themes and others Gnome customizations

### 3. Clean dependencies
Clean packages with `sudo apt -f -y install` to fix broken packages and `sudo apt-get autoremove && sudo apt-get clean` to clean.
### 4. Download my dotfiles
Download my [Dotfiles repository](https://github.com/frankjuniorr/dotfiles), but not install, only download. Because of my dotfiles installation is necessary the user interation.

## Install without git in a Ubuntu formatted
```bash
wget -O - https://git.io/fjqyx | bash
```
This shorten link ``https://git.io/fjqyx`` is pointed to ``https://raw.githubusercontent.com/frankjuniorr/ubuntu_install/master/code/trigger.sh``

# Screenshots after installation

#### Theme:

- [Adapta GTK Theme](https://github.com/adapta-project/adapta-gtk-theme)
- [Korla Icon Theme](https://github.com/bikass/korla)

#### Extensions:

- [Hide Activities](https://extensions.gnome.org/extension/1128/hide-activities-button/)
- [Move Clock](https://extensions.gnome.org/extension/2/move-clock/)
- [OpenWeather](https://extensions.gnome.org/extension/750/openweather/)


### Desktop:

<img alt="Icon" src="screenshot/screenshot_desktop.png?raw=true" align="center" hspace="1" vspace="1">

### Lock Screen:

* The image is same wallpaper, but is blur

<img alt="Icon" src="screenshot/screenshot_lockscreen.png?raw=true" align="center" hspace="1" vspace="1">

### Login Screen:

* The ``ubuntu_install`` set my user profile picture too =D

<img alt="Icon" src="screenshot/screenshot_login.png?raw=true" align="center" hspace="1" vspace="1">


CHANGELOG.md

v1.0 / 2019-04-18
=================

  * Merge pull request #6 from frankjuniorr/dev
  * update workflow
  * update README
  * remove debug
  * removedebug and improve profile image file
  * add some debugs
  * add some debugs
  * fix the feature of replace user profile picture
  * fix the feature of replace user profile picture
  * fix the feature of replace user profile picture
  * add comment
  * add profile picture, xclip, korla theme, and vs code extension
  * update README
  * change directory structure
  * disable warnings and gnome extensions
  * disable warnings and gnome extensions
  * add blur image in the lockscreen wallpaper
  * url de teste
  * url de teste
  * url de teste
  * remove sublime and atom, and add VS Code installation
  * change action workflow
  * change action workflow
  * Merge pull request #5 from frankjuniorr/dev
  * add my action
  * remove atom and sublime installation

master / 2019-03-28
===================

  * update screenshot
  * remove input user vars
  * installing dofiles automatically
  * installing dofiles automatically
  * add git email in prompt_vars
  * add git email in prompt_vars
  * add git email in prompt_vars
  * add git email in prompt_vars
  * fix bug with env vars
  * debug env vars
  * debug env vars
  * fix bug with env vars
  * fix bug with env vars
  * improve gnome extension installation
  * improve gnome extension installation
  * refactor gnome_extensions install
  * remove Calibre of main installation
  * debug gnome extensions
  * remove ansible.env
  * instaling Gnome Extensions
  * add some ubuntu jobs
  * add some ubuntu jobs
  * change Telegram installation directory
  * testing ansible more verbose
  * add another gsetting job
  * add gnome-shell-extensions in the list of packages
  * replace shell to gsettings ansible module
  * Update README.md
  * disable desktop icons
  * add more packages
  * Update README.md
  * add a screenshot example
  * Update README.md
  * Update README.md
  * melhoria no trigger
  * add vim in installation
  * add my dotfiles
  * add more gnome customization
  * add more gnome customization
  * add comments and improve Telegram installation
  * add gnome favorites
  * update gnome customize install
  * add wallpaper
  * [#4] Implement Sublime install
  * set Adapta theme by default
  * set Adapta theme by default
  * add a albert on boot
  * fix terminator installation
  * update README
  * add a albert on boot
  * Update README
  * add a trigger script
  * add Terminator installation
  * turn off warning in 2 commands
  * update comments
  * delete old file
  * add a finish message
  * fix dropbox installation
  * fix telegram installation
  * fix telegram installation
  * fix albert installation
  * fix albert installation
  * fix albert installations
  * add installation by Ansible
  * substituído screenfetch por neofetch
  * add screenfecth
  * adicionando 'pigments' Atom package
  * instalando o minimap e outros Atom packages para highlight
  * instalando o minimap e outros Atom packages para highlight
  * Merge branch 'master' of github.com:frankjuniorr/ubuntu_install
  * add Android tools adb package installation
  * Update README.md
  * added 'jq' installation
  * Merge branch 'master' of github.com:frankjuniorr/ubuntu_install
  * add TODO and fix some bugs and improvments
  * Merge pull request #2 from familqasimov/master
  * Update README.md
  * Add files via upload
  * Update README.md
  * update README
  * add a ISSUE_TEMPLATE and refactor code to a new directory
  * first commit
  * Initial commit


code/ubuntu_install.sh
#!/bin/bash

################################################################################
# Descrição:
#   Script to initial configs to Ubuntu 18.04 Minimal Installation.
#
################################################################################
# Uso:
#    ./ubuntu_install.sh
#
################################################################################
# Autor: Frank Junior <frankcbjunior@gmail.com>
# Desde: 18-10-2017
# Versão: 1
################################################################################


################################################################################
# Configurações
# set:
# -e: se encontrar algum erro, termina a execução imediatamente
set -e

################################################################################
# Variáveis - todas as variáveis ficam aqui

baseDir=$(pwd)

################################################################################
# Utils - funções de utilidades

# códigos de retorno
# [condig-style] constantes devem começar com 'readonly'
readonly SUCESSO=0
readonly ERRO=1

# debug = 0, desligado
# debug = 1, ligado
debug=0

# ============================================
# Função pra imprimir informação
# ============================================
_print_info(){
  local amarelo="\033[33m"
  local reset="\033[m"

  printf "${amarelo}$1${reset}\n"
}

# ============================================
# Função pra imprimir mensagem de sucesso
# ============================================
_print_success(){
  local verde="\033[32m"
  local reset="\033[m"

  printf "${verde}$1${reset}\n"
}

# ============================================
# Função pra imprimir mensagem de titulo
# ============================================
_print_title(){
  local text_cyan="$(tput setaf 6 2>/dev/null || echo '\e[0;36m')"
  local text_reset="$(tput sgr 0 2>/dev/null || echo '\e[0m')"

  printf "${text_cyan}$1${text_reset}\n"
}

# ============================================
# Função pra imprimir erros
# ============================================
_print_error(){
  local vermelho="\033[31m"
  local reset="\033[m"

  printf "${vermelho}[ERROR] $1${reset}\n"
}

# ============================================
# Função de debug
# ============================================
_debug_log(){
  [ "$debug" = 1 ] && _print_info "[DEBUG] $*"
}

# ============================================
# tratamento das exceções de interrupções
# ============================================
_exception(){
  return "$ERRO"
}

################################################################################
# Validações - regras de negocio até parametros

# ============================================
# tratamento de validacoes
# ============================================
validacoes(){
	return "$SUCESSO"
}

################################################################################
# Funções do Script - funções próprias e específicas do script

# ============================================
# Fazendo as atualizações iniciais
# ============================================
init_updates(){
  # updagrade inicial, por volta de uns 300 MB
  _print_info "Fazendo as atualizações iniciais..."
  sudo apt -y upgrade
  sudo apt -y dist-upgrade
  sudo apt -y full-upgrade

  # instalações dos pacotes de comunicação primeiro.
  sudo apt install -y curl wget
}

# ============================================
# adicionando repositórios dos Themes
# ============================================
add_themes(){
  _print_info "Adicionando repositórios dos Themes..."

  # para o Adapta theme (Material Design)
  if [ ! -f /etc/apt/sources.list.d/tista-ubuntu-adapta-bionic.list ];then
    sudo apt-add-repository ppa:tista/adapta -y
  else
    _print_success "Repositório do Adapta-theme já existe"
  fi

  if [ ! -f /etc/apt/sources.list.d/papirus-ubuntu-papirus-bionic.list ];then
    # pacote de icones pro Adapta theme (icones Material)
    sudo apt-add-repository ppa:papirus/papirus -y
  else
    _print_success "Repositório do Adapta-icon (Papirus) já existe"
  fi

  if [ ! -f /etc/apt/sources.list.d/communitheme-ubuntu-ppa-bionic.list ];then
    # instalação do CommunityTheme
    sudo add-apt-repository ppa:communitheme/ppa -y
  else
    _print_success "Repositório do CommunitTheme já existe"
  fi

}

# ============================================
# adicionando repositório do Spotify
# ============================================
add_spotify_repo(){
  if [ ! -f /etc/apt/sources.list.d/spotify.list ];then
    _print_info "Adicionando repositório do Spotify..."
    # 1. Add the Spotify repository signing keys to be able to verify downloaded packages
    sudo apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv-keys 0DF731E45CE24F27EEEB1450EFDC8610341D9410
    # 2. Add the Spotify repository
    echo deb http://repository.spotify.com stable non-free | sudo tee /etc/apt/sources.list.d/spotify.list
  else
    _print_success "Repositório do Spotify já existe"
  fi
}

# ============================================
# adicionando repositório do Atom
# ============================================
add_atom_repo(){
  if [ ! -f /etc/apt/sources.list.d/atom.list ];then
    _print_info "Adicionando repositório do Atom..."
    curl -L https://packagecloud.io/AtomEditor/atom/gpgkey | sudo apt-key add -
    sudo sh -c 'echo "deb [arch=amd64] https://packagecloud.io/AtomEditor/atom/any/ any main" > /etc/apt/sources.list.d/atom.list'
  else
    _print_success "repostiório do Atom já existe"
  fi
}

# ============================================
# adicionando repositório do LibreOffice
# ============================================
add_libreOffice_repo(){
  if [ ! -f /etc/apt/sources.list.d/libreoffice-ubuntu-ppa-bionic.list ];then
    _print_info "Adicionando repositório do LibreOffice..."
    sudo add-apt-repository ppa:libreoffice/ppa -y
  else
    _print_success "o repositório do LibreOffice já está instalado"
  fi
}

# ============================================
# adicionando repositório do Transmission
# ============================================
add_transmission_repo(){
  if [ ! -f /etc/apt/sources.list.d/transmissionbt-ubuntu-ppa-bionic.list ];then
    _print_info "Adicionando repositório do Transmission..."
    sudo add-apt-repository ppa:transmissionbt/ppa -y
  else
    _print_success "repositório do Transmission já existe"
  fi
}

# ============================================
# instalação dos Dark Themes
# ============================================
install_themes(){

  _print_info "Instalando temas..."

  sudo apt -y install \
  adapta-gtk-theme \
  papirus-icon-theme \
  ubuntu-communitheme-session
}

# ============================================
# instalação dos Calibre - Kindle ebook management
# ============================================
install_calibre(){
  if ! type calibre > /dev/null 2>&1; then
    _print_info "Instalando o Calibre..."
    sudo -v && wget -nv -O- https://download.calibre-ebook.com/linux-installer.sh | sudo sh /dev/stdin
  else
    _print_success "O Calibre já está instalado"
  fi
}

# ============================================
# instalação do Google Chrome
# ============================================
install_google_chrome(){
  if ! type google-chrome > /dev/null 2>&1; then
    _print_info "Instalando Google Chrome..."
    # fazendo download
    wget https://dl.google.com/linux/direct/google-chrome-stable_current_amd64.deb
    # instalando usando o 'gdebi'
    sudo gdebi google-chrome-stable_current_amd64.deb
    # removendo o pacote.deb
    rm google-chrome-stable_current_amd64.deb
  else
    _print_success "O Google Chrome já está instalado"
  fi
}

# ============================================
# Instala o Telegram Desktop
# ============================================
install_telegram(){
  cd ~ > /dev/null

  if [ ! -d $HOME/Telegram ];then
    _print_info "Instalando Telegram Desktop..."
    # fazendo donwload do Telegram
    wget https://telegram.org/dl/desktop/linux -O telegram.tar.xz
    wget https://telegram.org/img/t_logo.png -O telegram_logo_raw.png
    # extraindo o arquivo compactado
    tar -xvf telegram.tar.xz

    # convertendo o tamanho do icone para 128x128 (o original vem (256x256)
    convert -resize 128x128 telegram_logo_raw.png telegram_logo.png

    mv telegram_logo.png Telegram/

    # deletando os arquivos
    rm telegram.tar.xz
    rm telegram_logo_raw.png

    local telegramDesktopFile='[Desktop Entry]
    Type=Application
    Exec=/home/@user@/Telegram/Telegram
    Icon=/home/@user@/Telegram/telegram_logo.png
    Hidden=false
    NoDisplay=false
    Name[en_US]=Telegram
    Name=Telegram
    Comment[en_US]=
    Comment=
    '

    telegramDesktopFile=$(echo "$telegramDesktopFile" | sed "s/@user@/$USER/g")

    echo "$telegramDesktopFile" > $HOME/.local/share/applications/telegram.desktop

    _print_success "O Telegram foi instalado em $HOME/Telegram"
  else
    _print_success "Telegram já está instalado em $HOME/Telegram"
  fi

  cd "$baseDir" > /dev/null

}

# ============================================
# Função genérica que instala pacotes do Atom.
# $1 - nome do pacote
# ============================================
_install_atom_package(){
  local package_name="$1"

  if [ ! -d "$HOME/.atom/packages/$package_name" ];then
    apm install "$package_name"
  else
    _print_success "$package_name já está instalado"
  fi
}

# ============================================
# Instala plugins and add-ons pro Atom.
# ============================================
upgrade_atom(){
  _print_info "Upgrade Atom..."

  packages=(
  "file-icons"
  "atom-material-ui"
  "atom-material-syntax-dark"
  "atom-beautify"
  "language-docker"
  "language-conky"
  "pigments"
  "minimap")

  for package in "${packages[@]}";s do
    _install_atom_package "$package"
  done

  # Python code that use to formatting Shell Script.
  # this is necessary to [atom-beautify] works
  # pip install beautysh
}

# ============================================
# instalando o Dropbox
# ============================================
install_dropbox(){

  if [ ! -d ~/.dropbox-dist/ ];then
    _print_info "Instalando o Dropbox..."

    cd ~ && wget -O - "https://www.dropbox.com/download?plat=lnx.x86_64" | tar xzf -

    dropboxFile='[Desktop Entry]
    Type=Application
    Exec=/home/@user@/.dropbox-dist/dropboxd
    Hidden=false
    NoDisplay=false
    X-GNOME-Autostart-enabled=true
    Name[en_US]=Dropbox
    Name=Dropbox
    Comment[en_US]=
    Comment=
    '
    dropboxFile=$(echo "$dropboxFile" | sed "s/@user@/$USER/")

    if [ ! -d ~/.config/autostart ];then
      mkdir ~/.config/autostart
    fi
    echo "$dropboxFile" > ~/.config/autostart/dropbox.desktop
  else
    _print_success "Dropbox já está instalado"
  fi

  cd "$baseDir" > /dev/null
}

# ============================================
# instala ferramentas e programas utils
# ============================================
install_packages(){
  _print_info "Instalando algumas ferramentas e utilitários..."

  # [gksu] - (DEU ERRO) para executar o “Disk Usage Analyzer (baobab)”, usando o  “gksudo baobab”
  # [git]
  # [meld] - usado para os diffs do git
  # [vim]
  # [jq] - leitor de JSON
  # [gdebi-core] - 'gdebi' is a simple way to install deb files, installaling dependencies too
  # [transmission] - torrent client
  # [libreOffice] - packages about LibreOffice
  # [gnome-tweak-tool] - usado para customizar a interface gráfica
  # [atom] - IDE
  # [python-pip] - Instalador de pacotes do Python
  # [browser-plugin-vlc] - VLC
  # [lame] - create mp3 audio files
  # [sound-juicer] - rip cdrom
  # [ImageMagick] - pacote para manipualções de imagens
  # [ubuntu-restricted-extras] - pacotes extras do Ubuntu: mp3 codec, font tts da Microsoft....
  # [neofetch] - Imprime system logo e sysinfo

  sudo apt -y install \
  git \
  meld \
  vim \
  jq \
  gdebi-core \
  transmission \
  libreoffice-gtk2 libreoffice-gnome \
  gnome-tweak-tool \
  spotify-client \
  atom \
  python-pip \
  browser-plugin-vlc \
  lame \
  sound-juicer \
  imagemagick \
  ubuntu-restricted-extras \
  neofetch

  install_telegram
  install_google_chrome
  install_calibre
  install_dropbox

  upgrade_atom
}

# ============================================
# limpa dependencias
# ============================================
clean_dependencies(){
  # apt fix-broken
  # pra consertar possíveis dependencias e pacotes quebrados
  sudo apt -f -y install
  # remover pacotes não mais utilizados
  sudo apt-get autoremove -y
  # deletar os pacotes .deb em '/var/cache/apt/archives/'
  sudo apt-get clean -y
}

# ============================================
# mostrar o banner inicial
# ============================================
show_header(){
cat << "HEADER"
   __  ____                __           ____           __        ____
  / / / / /_  __  ______  / /___  __   /  _/___  _____/ /_____ _/ / /
 / / / / __ \/ / / / __ \/ __/ / / /   / // __ \/ ___/ __/ __ `/ / /
/ /_/ / /_/ / /_/ / / / / /_/ /_/ /  _/ // / / (__  ) /_/ /_/ / / /
\____/_.___/\__,_/_/ /_/\__/\__,_/  /___/_/ /_/____/\__/\__,_/_/_/

HEADER
}

# ============================================
# Função Main
# ============================================
main(){

  _print_title "==========================================="
  _print_title "[01/06] Upgrade inicial..."
  _print_title "==========================================="
  echo ""

  init_updates

  _print_title "==========================================="
  _print_title "[02/06] Add repositories..."
  _print_title "==========================================="
  echo ""

  add_spotify_repo
  add_atom_repo
  add_themes
  add_libreOffice_repo
  add_transmission_repo

  # Atualizando....
  _print_title "==========================================="
  _print_title "[03/06] Update packages..."
  _print_title "==========================================="
  echo ""

  sudo apt update

  _print_title "==========================================="
  _print_title "[04/06] Install packages..."
  _print_title "==========================================="
  echo ""

  install_themes
  install_packages

  _print_title "==========================================="
  _print_title "[05/06] Clean dependencies..."
  _print_title "==========================================="
  echo ""

  clean_dependencies

  _print_title "==========================================="
  _print_title "[06/06] Install development environment..."
  _print_title "==========================================="
  echo ""

  cd "$baseDir" > /dev/null
  ./development_install.sh

  clear
  _print_success "pronto, tudo OK."
  _print_success "Se você acabou de formatar o computador, é bom resetar a máquina =D"
}

################################################################################
# Main - execução do script

# trata interrrupção do script em casos de ctrl + c (SIGINT) e kill (SIGTERM)
trap _exception SIGINT SIGTERM

validacoes
show_header
main

################################################################################
# FIM do Script =D
